cpp/src/neighbors/detail/vamana/priority_queue.cuh

```diff
@@ -0,0 +1,382 @@
+
+
+    // Pop root node off and heapify
+    __device__ DistPair<IdxT,accT> pop(){
+      DistPair<IdxT,accT> result;
+      result.dist = vals[0].dist;
+      result.idx = vals[0].idx;
+      vals[0].dist =  INFTY<accT>();
+       vals[0].idx = INFTY<IdxT>();
+       heapify();    
+       *q_size -= 1;         
+       return result;
+    }
+
+};
+
+
+/***************************************************************************************
+* Node structure used for simplified lists during GreedySearch.
+* Used for other operations like checking for duplicates, etc.
+****************************************************************************************/
+template <typename SUMTYPE> class __align__(16) Node {
+public:
+  SUMTYPE distance;
+  int nodeid;
+};
+
+// Less-than operator between two Nodes.
+template <typename SUMTYPE>
+__host__ __device__ bool operator<(const Node<SUMTYPE> &first,
+                                   const Node<SUMTYPE> &other) {
+  return first.distance < other.distance;
+}
+
+// Less-than operator between two Nodes.
+template <typename SUMTYPE>
+__host__ __device__ bool operator>(const Node<SUMTYPE> &first,
+                                   const Node<SUMTYPE> &other) {
+  return first.distance > other.distance;
+}
+
+template <typename accT>
+__device__ bool check_duplicate(const Node<accT> *pq, const int size,
+                                 Node<accT> new_node) {
+
+  bool found = false;
+  for (int i = threadIdx.x; i < size; i += blockDim.x) {
+    if (pq[i].nodeid == new_node.nodeid) {
+      found = true;
+      break;
+    }
+  }
+
+  unsigned mask = __ballot_sync(FULL_BITMASK, found);
+
+  if (mask == 0)
+    return false;
+
+  else
+    return true;
+}
+
+/*  
+  Enqueing a input value into parallel queue with tracker
+*/
+template <typename SUMTYPE>
+__inline__ __device__ void
+parallel_pq_max_enqueue(Node<SUMTYPE> *pq, int *size, const int pq_size,
+                        Node<SUMTYPE> input_data, SUMTYPE *cur_max_val, int* max_idx) {
+
+  if (*size < pq_size) {
+    __syncthreads();
+    if (threadIdx.x == 0) {
+      pq[*size].distance = input_data.distance;
+      pq[*size].nodeid = input_data.nodeid;
+      *size = *size + 1;
+      if (input_data.distance > (*cur_max_val)){
+        *cur_max_val = input_data.distance;
+        *max_idx = *size - 1;
+      }
+    }
+    __syncthreads();
+    return;
+  }
+  else {
+
+    if (input_data.distance >= (*cur_max_val)) {
+      __syncthreads();
+      return;
+    }
+    if(threadIdx.x == 0) {
+      pq[*max_idx].distance = input_data.distance;
+      pq[*max_idx].nodeid = input_data.nodeid;
+    }
+    int idx = 0;
+    SUMTYPE max_val = pq[0].distance;
+
+    for (int i = threadIdx.x; i < pq_size; i += 32) {
+      if (pq[i].distance > max_val) {
+        max_val = pq[i].distance;
+        idx = i;
+      }
+    }
+
+    for (int offset = 16; offset > 0; offset /= 2) {
+      SUMTYPE new_max_val = __shfl_down_sync(FULL_BITMASK, max_val, offset);
```

Comment:
Moved most of the calls to their raft versions, with the exception of the distance metric calls in `vamana_structs.cuh`, because the logic and optimizations are a bit more complex there. Need to convert to raft calls and then verify correctness and no perf impact.